src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,362 @@
+//! Memory allocation APIs
+
+#![stable(feature = "alloc_module", since = "1.28.0")]
+
+mod global;
+mod layout;
+
+#[stable(feature = "global_alloc", since = "1.28.0")]
+pub use self::global::GlobalAlloc;
+#[stable(feature = "alloc_layout", since = "1.28.0")]
+pub use self::layout::{Layout, LayoutErr};
+
+use crate::fmt;
+use crate::ptr::{self, NonNull};
+
+/// The `AllocErr` error indicates an allocation failure
+/// that may be due to resource exhaustion or to
+/// something wrong when combining the given input arguments with this
+/// allocator.
+#[unstable(feature = "allocator_api", issue = "32838")]
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct AllocErr;
+
+// (we need this for downstream impl of trait Error)
+#[unstable(feature = "allocator_api", issue = "32838")]
+impl fmt::Display for AllocErr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("memory allocation failed")
+    }
+}
+
+/// A desired initial state for allocated memory.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[unstable(feature = "allocator_api", issue = "32838")]
+pub enum AllocInit {
+    /// The contents of the new memory are undefined.
+    ///
+    /// Reading uninitialized memory is Undefined Behavior; it must be initialized before use.
```

Comment:
This is not true in its full generality. For example, `ptr::read::<MaybeUninit<u32>>` is completely legal on uninitialized memory.
See [the UB docs](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) for more details.

Reply:
I'd say we remove the section on undefined behavior, as also reading zeroed-memory may be undefined behavior (i.e. reading into `&u32`).

Reply:
Makes sense.
Also, instead of saying the contents are "undefined" I think it is better to say they are "uninitialized" (which is a precisely defined-state for memory to be in in the Rust Abstract Machine).

Reply:
Fixed in #70691